api/helpers.go

```diff
@@ -66,3 +67,40 @@ func (a *API) requestAud(ctx context.Context, r *http.Request) string {
 	// Finally, return the default of none of the above methods are successful
 	return config.JWT.Aud
 }
+
+func (a *API) isEmailBlacklisted(email string) bool {
+	globalConfig := a.config
+	if a.blacklist.UpdateNeeded() {
+		a.blacklist.UpdateFromURL(globalConfig.EmailBlacklist.URL)
+	}
+	return a.blacklist.EmailBlacklisted(email)
+}
+
+func (a *API) isPasswordValid(password string) bool {
+	re_min_nums := regexp.MustCompile("[0-9]")
+	re_min_symbols := regexp.MustCompile("[$@$!%*#?&]")
+	re_min_uppercase := regexp.MustCompile("[A-Z]")
+
+	switch {
+	case a.config.Password.MinNumbers > 0:
+		if len(re_min_nums.FindAllString(password, -1)) < a.config.Password.MinNumbers {
+			return false
+		}
+		fallthrough
+	case a.config.Password.MinSymbols > 0:
+		if len(re_min_symbols.FindAllString(password, -1)) < a.config.Password.MinSymbols {
+			return false
+		}
+		fallthrough
+	case a.config.Password.MinUppercase > 0:
+		if len(re_min_uppercase.FindAllString(password, -1)) < a.config.Password.MinUppercase {
+			return false
+		}
+		fallthrough
+	default:
+		if l := len(password); l == 0 || l < a.config.Password.MinLength {
+			return false
+		}
+		return true
```

Comment:
The use of a switch here is kind of odd to me. If you're doing a fallthrough I think just an if ladder would be clearer. 

- having the length check up top would be good, it is the cheapest check. 

- returning a message would be something to consider, a reason that the password failed ("Passwords must be over XXX characters"). In that case, you might consider coalescing the errors, but it isn't necessary.

Reply:
@rybit I like the idea of coalescing errors, but that would require every check to run. Otherwise we could:

- Return a description of the first error that caused the password validation to fail
- Create a generic error message to return that contains all validation requirements (i.e. `password requires 1 number, 2 uppercase letters, and 3 symbols`)
  * This could be overly verbose

Reply:
you don't have to run each check, similar to how you have it structured you'd just return the errors/string at the end.

``` go
errors = []string{}
if [some condition failing] {
  errors = append(errors, fmt.Sprintf("must do this thing %d times", times))
}

.... rinse and repeat

if len(errors) == 0 {
  return "", true
}
return strings.Join(errors, ","), false
```

or something like that. And, yeah it could def get verbose.